### PR TITLE
rbd: Add rbd_sparsify_with_progress rbd API

### PR DESCRIPTION
**Pull request by Nikhil-Ladha**
**Description:**
Added a new rbd API `rbd_sparsify_with_progress`, and supporting tests.

Fixes: #281

## Checklist
- [X] Added tests for features and functional changes
- [X] Public functions and types are documented
- [X] Standard formatting is applied to Go code
- [X] Is this a new API? Added a new file that begins with `//go:build ceph_preview`
- [X] Ran `make api-update` to record new APIs

New or infrequent contributors may want to review the go-ceph [Developer's Guide](https://github.com/ceph/go-ceph/blob/master/docs/development.md) including the section on how we track [API Status](https://github.com/ceph/go-ceph/blob/master/docs/development.md#api-status) and the [API Stability Plan](https://github.com/ceph/go-ceph/blob/master/docs/api-stability.md).

The go-ceph project uses mergify. View the [mergify command guide](https://docs.mergify.com/commands/#commands) for information on how to interact with mergify. Add a comment with `@Mergifyio` `rebase` to rebase your PR when github indicates that the PR is out of date with the base branch.

### CODE REVIEW

File: rbd/rbd_sparsify_test.go
Comment (anoopcs9) [inline rbd/rbd_sparsify_test.go:1]: The test file is missing `ceph_preview` build tag as in the above source file.

File: rbd/rbd_sparsify.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
+//go:build ceph_preview
+// +build ceph_preview
+
+package rbd
+
+/*
+#cgo LDFLAGS: -lrbd
+#include <errno.h>
```
Comment (phlogistonjohn) [inline rbd/rbd_sparsify.go:8]: Double check: what was errno.h needed for?

File: rbd/rbd_sparsify.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+#include <stdlib.h>
+#include <rbd/librbd.h>
+#include <rados/librados.h>
```
Comment (phlogistonjohn) [inline rbd/rbd_sparsify.go:11]: Double check: what was librados.h needed for?

File: rbd/rbd_sparsify.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+type sparsifyCallbackCtx struct {
+	callback SparsifyCallback
+	data     interface{}
+}
+
+// SparsifyWithProgress makes an image sparse by deallocating runs of zeros.
+// The sparseSize value will be used to find runs of zeros and must be
+// a power of two no less than 4096 and no larger than the image size.
+// The given progress callback will be called to report on the progress
+// of sparse.
```
Comment (phlogistonjohn) [inline rbd/rbd_sparsify.go:46]: IIRC, returning a non-zero value from the callback will also abort the function. It may be good to (a) verify that I'm correct and if so (b) mention it in the doc comment.

File: rbd/rbd_sparsify.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+//
+// Implements:
+//
+//	int rbd_sparsify_with_progress(rbd_image_t image, size_t sparse_size,
+//								   librbd_progress_fn_t cb, void *cbdata);
+func (image *Image) SparsifyWithProgress(sparseSize uint, cb SparsifyCallback,
+	data interface{}) error {
```
Comment (phlogistonjohn) [inline rbd/rbd_sparsify.go:53]: Style nitpick: I find it easier to read a line-split set of arguments if the split occurs before the arguments. So something like (psueocode-ish):
```
func (image *Image) SparsifyWithProgress(
	sparseSize uint, cb SparsifyCallback, data interface{}) error {
```
but it's not a huge deal

File: rbd/rbd_sparsify_test.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+		assert.NoError(t, err)
+		defer func() { assert.NoError(t, img.Close()) }()
+
+		cc := 0
+		cb := func(offset, total uint64, v interface{}) int {
+			cc++
+			val := v.(int)
+			assert.Equal(t, 0, val)
+			assert.Equal(t, uint64(1), total)
+			return 0
```
Comment (phlogistonjohn) [inline rbd/rbd_sparsify_test.go:39]: For extra bonus points it might be good to add a test case that return non zero to abort the sparsify call (see my other comment). I won't demand it for approval though.

File: rbd/rbd_sparsify_test.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+	t.Run("invalidValue", func(t *testing.T) {
+		img, err := OpenImage(ioctx, name, NoSnapshot)
+		assert.NoError(t, err)
+		defer func() { assert.NoError(t, img.Close()) }()
+
+		err = img.SparsifyWithProgress(4096, nil, nil)
+		assert.Error(t, err)
+	})
+}
```
Comment (phlogistonjohn) [inline rbd/rbd_sparsify_test.go:55]: I think you might be able to achieve 100% code coverage if you also add a test for the invalid "image is closed" condition. That would force it to enter the if statement on line 59 of the new go code.

**Review by phlogistonjohn — CHANGES_REQUESTED** [review]
Nitpick:  Remove the text `**Description:**` from the commit message. It might look nice on github when it renders the commit in the PR description but if you are like me and mainly use git on the CLI, it doesn't really add anything to the commit.

File: rbd/rbd_sparsify_test.go
Comment (Nikhil-Ladha) [reply rbd/rbd_sparsify_test.go:1]: Fixed

File: rbd/rbd_sparsify.go (hunk 1 above)
Comment (Nikhil-Ladha) [reply rbd/rbd_sparsify.go:8]: The test was complaining about no known kind of `C.EINVAL` if the `errno.h` header was not included.

File: rbd/rbd_sparsify.go (hunk 2 above)
Comment (Nikhil-Ladha) [reply rbd/rbd_sparsify.go]: Not required, removed it.

File: rbd/rbd_sparsify.go (hunk 3 above)
Comment (Nikhil-Ladha) [reply rbd/rbd_sparsify.go]: Yeah, it does return a `EPERM` error.
Updated the doc.

File: rbd/rbd_sparsify.go (hunk 4 above)
Comment (Nikhil-Ladha) [reply rbd/rbd_sparsify.go]: No issues, updated the styling

File: rbd/rbd_sparsify_test.go (hunk 5 above)
Comment (Nikhil-Ladha) [reply rbd/rbd_sparsify_test.go:42]: Added an extra TC

File: rbd/rbd_sparsify_test.go (hunk 6 above)
Comment (Nikhil-Ladha) [reply rbd/rbd_sparsify_test.go:94]: Added an extra TC for closed image

**Review by phlogistonjohn — APPROVED** [review]
LGTM. Thanks for another great contribution!

File: rbd/rbd_sparsify.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+		image, sparse_size, (librbd_progress_fn_t)sparsifyCallback, (void*)arg);
+};
+*/
+import "C"
+
+import (
+	"github.com/ceph/go-ceph/internal/callbacks"
+)
+
+// SparsifyCallback defines the function signature needed for the
```
Comment (ansiwen) [inline rbd/rbd_sparsify.go:27]: We should document the semantics of the callback parameters here. (Or is it something super obvious that I don't get?) That would also include the return value, so better we document it here as well, instead of in the docs for `SparsifyWithProgress`.

Comment (Nikhil-Ladha) [reply rbd/rbd_sparsify.go:27]: When you say semantics, you mean an `implementation` line similar to the other function doc comments, right?
Or do you mean something else?

Comment (ansiwen) [reply rbd/rbd_sparsify.go:27]: @Nikhil-Ladha I just meant that I - as a user of this library - would have no idea what the parameters of the callback mean. On the other hand, you are right that we often just refer to the functions of the C API of the ceph libraries, which is also not very user-friendly, because I know how hard it is sometimes, to find the right documentation. But if it is worth mentioning the behavior of the return value, it should also be worth to document the semantics of the parameters. Actually I'm not sure if we have a clear pattern of what to document and what to refer to the upstream docs. @phlogistonjohn, you have an opinion or clearer rule in your mind?

Comment (phlogistonjohn) [reply rbd/rbd_sparsify.go:27]: I think in a case like this it is good to encourage putting in some docs that explain what the expected parameters are. Note that we didn't do it for group_snap.go, so I wouldn't demand it mysefl. That said,  I think that in this case it is fairly straightforward to add something based on the names of the argument to the librbd_progress_fn_t  (at https://github.com/ceph/ceph/blob/2dee65108fc1e4adfad36ac9bcb4fd35dd8f6420/src/include/rbd/librbd.h#L77 ). So something like `The callback function will be called with the first argument containing the current offset within the image being made sparse and the second argument containing the total size of the image. The third argument is an opaque value that is passed to the SparsifyWithProgress function's data argument and every call to the callback will receive the same object.`  Feel free to take, leave, or modify to your liking.

**Comment by anoopcs9** [comment]
@Mergifyio rebase

**Review by phlogistonjohn — APPROVED** [review]

**Review by ansiwen — APPROVED** [review]
Docs changes will be handled in a follow-up PR.

**Comment by phlogistonjohn** [comment]
> Docs changes will be handled in a follow-up PR.

See: https://github.com/ceph/go-ceph/pull/861
